<a href="https://colab.research.google.com/github/dwkaris/Week9Project/blob/master/DuncanKariuki_Week9_NaiveBayes_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#WEEK 9 INDEPENDENT PROJECT

##DEFINING THE ANALYTIC QUESTION QUESTION

###a). Specifying the analytic question

As Data Scientists, we have been tasked with building a model that correctly predict if an email is a spam or not

###b). Defining the metrics of success

Our success will be determined by building a model with the highest accuracy, we will use the confusion matrix and classification report to determie our accuracy.

###c) Understanding the Context

As Data Scientists, we have been tasked with studying data containing frequecy that a word appears collected from recieved Emails, classified as spam and not spam. Later onwards, build a model that predicts whether a recieved Email is spam or not.

###d). Recording the Experimental Design

Below is the expected workflow of our Analysis

1.   Data Cleaning and Preparation

*   Load libraries and dataset

2. Tidying up the dataset
  
3.  Spliting dataset into testing and training data

4.  Perform classification using the best Naive Bayes model.

5. Compute the accuracy (percentage of correct classification).

6.  Provide further recommendations to improve both classifiers.


###e) Data relevance/ Appropriateness of Data

Our study aims at creating the best model that predicts whether or not a recieved email is a Spam email.Below is the link:

1. https://archive.ics.uci.edu/ml/datasets/Spambase


## Importing Libraries

In [74]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 5000)
pd.set_option('display.width', 5000)
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import  StandardScaler
from sklearn import metrics
from sklearn import preprocessing
from sklearn.naive_bayes import MultinomialNB,GaussianNB,ComplementNB

## Load the Dataset

In [57]:
# Load Dataset

spam = pd.read_csv('spambasedata.csv', delimiter=',')


In [58]:
# Preview the top 5 rows of our dataset

spam.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  word_freq_order  word_freq_mail  word_freq_receive  word_freq_will  word_freq_people  word_freq_report  word_freq_addresses  word_freq_free  word_freq_business  word_freq_email  word_freq_you  word_freq_credit  word_freq_your  word_freq_font  word_freq_000  word_freq_money  word_freq_hp  word_freq_hpl  word_freq_george  word_freq_650  word_freq_lab  word_freq_labs  word_freq_telnet  word_freq_857  word_freq_data  word_freq_415  word_freq_85  word_freq_technology  word_freq_1999  word_freq_parts  word_freq_pm  word_freq_direct  word_freq_cs  word_freq_meeting  word_freq_original  word_freq_project  word_freq_re  word_freq_edu  word_freq_table  word_freq_conference  char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  capital_run_length_average  capital_run_length_longest  capital_run_length_total  class
0            0.00               0.64           0.64           0.0           0.32            0.00              0.00                0.00             0.00            0.00               0.00            0.64              0.00              0.00                 0.00            0.32                0.00             1.29           1.93              0.00            1.59             0.0           0.43             0.43           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0            0.07              0.0           0.0              0.00           0.0                0.0                0.00                0.0          0.00           0.00              0.0                   0.0         0.00        0.132          0.0        0.372        0.180        0.048                       5.114                       101.0                    1028.0    1.0
1            0.21               0.28           0.50           0.0           0.14            0.28              0.21                0.07             0.00            0.94               0.21            0.79              0.65              0.21                 0.14            0.14                0.07             0.28           3.47              0.00            0.51             0.0           1.16             0.06           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0            0.00              0.0           0.0              0.06           0.0                0.0                0.12                0.0          0.06           0.06              0.0                   0.0         0.01        0.143          0.0        0.276        0.184        0.010                       9.821                       485.0                    2259.0    1.0
2            0.06               0.00           0.71           0.0           1.23            0.19              0.19                0.12             0.64            0.25               0.38            0.45              0.12              0.00                 1.75            0.06                0.06             1.03           1.36              0.32            0.31             0.0           0.00             0.00           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0            0.00              0.0           0.0              0.00           0.0                0.0                0.00                0.0          0.00           0.00              0.0                   0.0         0.00        0.137          0.0        0.137        0.000        0.000                       3.537                        40.0                     191.0    1.0
3            0.00               0.00           0.00           0.0           0.63            0.00       

In [59]:
# Preview the last 5 rows of our 

spam.tail()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  word_freq_order  word_freq_mail  word_freq_receive  word_freq_will  word_freq_people  word_freq_report  word_freq_addresses  word_freq_free  word_freq_business  word_freq_email  word_freq_you  word_freq_credit  word_freq_your  word_freq_font  word_freq_000  word_freq_money  word_freq_hp  word_freq_hpl  word_freq_george  word_freq_650  word_freq_lab  word_freq_labs  word_freq_telnet  word_freq_857  word_freq_data  word_freq_415  word_freq_85  word_freq_technology  word_freq_1999  word_freq_parts  word_freq_pm  word_freq_direct  word_freq_cs  word_freq_meeting  word_freq_original  word_freq_project  word_freq_re  word_freq_edu  word_freq_table  word_freq_conference  char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  capital_run_length_average  capital_run_length_longest  capital_run_length_total  class
4596            0.31                0.0           0.62           0.0           0.00            0.31               0.0                 0.0              0.0             0.0                0.0            1.88              0.00               0.0                  0.0             0.0                 0.0              0.0           0.62               0.0            2.00             0.0            0.0              0.0           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0             0.0              0.0           0.0               0.0           0.0                0.0                 0.0               0.00          0.00           2.00              0.0                   0.0        0.000        0.000          0.0        0.353          0.0          0.0                       1.555                         4.0                      14.0    0.0
4597            0.00                0.0           0.00           0.0           0.00            0.00               0.0                 0.0              0.0             0.0                0.0            0.00              0.00               0.0                  0.0             0.0                 0.0              0.0           6.00               0.0            0.30             0.0            0.0              0.0           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0             0.0              0.0           0.0               0.0           0.0                0.0                 0.0               0.00          0.00           1.20              0.0                   0.0        0.102        0.718          0.0        0.000          0.0          0.0                       1.404                         6.0                     118.0    0.0
4598            0.30                0.0           0.30           0.0           0.00            0.00               0.0                 0.0              0.0             0.0                0.0            1.80              0.30               0.0                  0.0             0.0                 0.0              0.9           1.50               0.0            0.32             0.0            0.0              0.0           0.0            0.0               0.0            0.0            0.0             0.0               0.0            0.0             0.0            0.0           0.0                   0.0             0.0              0.0           0.0               0.0           0.0                0.0                 0.0               0.32          0.00           0.32              0.0                   0.0        0.000        0.057          0.0        0.000          0.0          0.0                       1.147                         5.0                      78.0    0.0
4599            0.96                0.0           0.00           0.0           0.32           

## Checking the data

In [60]:
# Checking the shape of our data
spam.shape

# our data has 4601 rows and 58 column

(4601, 58)

In [61]:
# Checking for infomation about a dataset

spam.info()

# our dataset has one null values
# all columns have numerical data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

## External Data Source Validation

Making sure your data matches something outside of the dataset is very important. It allows you to ensure that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset. External validation can often be as simple as checking your data against a single number, as we will do here.

## Tidying The Dataset

In [62]:
# We start by checking for null values in our dataset

spam.isnull().sum()

# Our dataset has only one null value hence we will drop it

spam.dropna(inplace=True)

In [63]:
# Check for unique values in our target column

spam['class'].value_counts()

# 2788 not spam emails
# 1812 spam emails

0.0    2788
1.0    1812
Name: class, dtype: int64

In [64]:
# Remove white spaces
spam.columns = spam.columns.str.strip().str.replace('','')
column = ['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d', 'word_freq_our',
          'word_freq_over', 'word_freq_remove', 'word_freq_internet', 'word_freq_order', 'word_freq_mail', 
          'word_freq_receive', 'word_freq_will', 'word_freq_people', 'word_freq_report', 'word_freq_addresses',
          'word_freq_free', 'word_freq_business', 'word_freq_email', 'word_freq_you', 'word_freq_credit',
          'word_freq_your', 'word_freq_font', 'word_freq_000', 'word_freq_money', 'word_freq_hp',
          'word_freq_hpl', 'word_freq_george', 'word_freq_650', 'word_freq_lab', 'word_freq_labs', 
          'word_freq_telnet', 'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85', 
          'word_freq_technology', 'word_freq_1999', 'word_freq_parts', 'word_freq_pm', 'word_freq_direct', 
          'word_freq_cs', 'word_freq_meeting', 'word_freq_original', 'word_freq_project', 'word_freq_re', 
          'word_freq_edu', 'word_freq_table', 'word_freq_conference', 'char_freq_;', 'char_freq_(',
          'char_freq_[', 'char_freq_!', 'char_freq_$', 'char_freq_#', 'capital_run_length_average',
          'capital_run_length_longest', 'capital_run_length_total']

for i, feat in enumerate(column):
  spam[feat] = spam[feat].replace('','')

## Implimenting the solution

## Using Multinomial Naive Bayes

###spliting our data into features and targets

In [65]:
# Split data into Features and Labels

X = spam.iloc[:, :-1].values
y = spam.iloc[:, -1:].values


### Split data into test  and train sets

In [66]:
# Split data into test and train sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 50)

### Initialize the Model

In [67]:
mnb = MultinomialNB()

mnb.fit(X_train, y_train)


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

### Predict X_test

In [68]:
y_pred = mnb.predict(X_test)

### Check accuracy of our model

In [69]:
print(metrics.classification_report(y_test, y_pred))

print(metrics.confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.83      0.83      0.83       576
         1.0       0.71      0.71      0.71       344

    accuracy                           0.78       920
   macro avg       0.77      0.77      0.77       920
weighted avg       0.78      0.78      0.78       920

[[476 100]
 [100 244]]


## Using Gausian Naive Bayes

In [70]:
# Split data into Features and Labels

X = spam.iloc[:, :-1].values
y = spam.iloc[:, -1:].values


# Split data into test and train sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 50)



###Fiting the model

In [71]:
gnb = GaussianNB()

gnb.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

### Predict X_test

In [72]:
y_pred = gnb.predict(X_test)

### Check accuracy of our model

In [73]:
print(metrics.classification_report(y_test, y_pred))

print(metrics.confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.96      0.72      0.82       576
         1.0       0.67      0.96      0.79       344

    accuracy                           0.81       920
   macro avg       0.82      0.84      0.80       920
weighted avg       0.85      0.81      0.81       920

[[412 164]
 [ 15 329]]


## Using the Complementary Distribution

In [75]:
# Split data into Features and Labels

X = spam.iloc[:, :-1].values
y = spam.iloc[:, -1:].values


# Split data into test and train sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 50)

# Fiting the model

cmb = ComplementNB()
cmb.fit(X_train, y_train)

# Predict X_test
y_pred = cmb.predict(X_test)

# Check accuracy of our model

print(metrics.classification_report(y_test, y_pred))

print(metrics.confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.83      0.82      0.83       576
         1.0       0.71      0.73      0.72       344

    accuracy                           0.78       920
   macro avg       0.77      0.77      0.77       920
weighted avg       0.79      0.78      0.79       920

[[472 104]
 [ 94 250]]


##Challenging the solution

The easy solution is nice because it is, well, easy, but you should never allow those results to hold the day. You should always be thinking of ways to challenge the results, especially if those results comport with your prior expectation.

From our analysis the data had already been preprocessed hence cuting our work to only fiting the best model

since the data contained the frequency of words and not the count, the Gausian dististribution produced the highest accuracy of 0.81 ie 81%. 412 correct not spam emails and 164 incorrect not spam emails, 329 correct spam emails and 15 incorrect spam emails.

Since the data had been preprocessed for us it was difficult to know if if it was preprossed correctly which may have led to inaccuracy of our data.


##Follow up questions

At this point, we can refine our question or collect new data, all in an iterative process to get at the truth.

a). Did we have the right data?

Yes
 
because using the data we were able to conduct an analysis on our analytic question

b). Do we need other data to answer our question?

no.

The data provided was sufficient

c). Did we have the right question?

yes

We were able to make conclusion from our data